### PR TITLE
Flipping `demand`

### PR DESCRIPTION
**Pull request by Anton-Latukha**
A milestone stage towards #850

Remember that `force` would lose the functional argument.
`demand` would be something along the lines of `free -> f =<< force thunk`.

By the result of the `demand` - it would be seen, is in viable to keep it, or simply replace with that `free -> f =<< force thunk` or move `free` logic into `force` and use `force` with `=<<`.

So instead of the `demand` it would be between function and its argument would be `force v`.

In other words, unflipping required to unfold the code.

But the further following folded elegantness worth it.

Also all those `\case` really are helper functions that explain themself.

Or, `alg` if you want.

### CODE REVIEW

**Comment by Anton-Latukha** [comment]
Currently surprised the benchmark saves are still small.

Sadly, the benchmarks so far not show huge improvement.

But need to inspect the profiling more.

I wait on one at the end of #850.

But more and more thinking that 1-threadness with no system of lock management, flipped style in which the project was created produces unoptimized code and lazy boxed types are the main points. Maybe they are the main nails performance hangs on in the thunk processing.

The more we streamline the code - the more magnitude of the gains would start to snowball.

**Comment by Anton-Latukha** [comment]
Awakened the swarm nest for project improvements:

#864
#861
#850
#776

, and others.

Checked for it to eval Nixpkgs in the same way. During evaluations found several typos, until evaluations became same.
